lib/class-wp-theme-json-gutenberg.php

```diff
@@ -141,77 +141,69 @@ class WP_Theme_JSON_Gutenberg {
 	 *
 	 * This contains the necessary metadata to process them:
 	 *
-	 * - path          => where to find the preset within the settings section
-	 *
-	 * - value_key     => the key that represents the value
-	 *
-	 * - value_func    => optionally, instead of value_key, a function to generate
-	 *                    the value that takes a preset as an argument
-	 *
-	 * - css_var_infix => infix to use in generating the CSS Custom Property. Example:
-	 *                   --wp--preset--<preset_infix>--<slug>: <preset_value>
-	 *
-	 * - classes      => array containing a structure with the classes to
-	 *                   generate for the presets. Each class should have
-	 *                   the class suffix and the property name. Example:
-	 *
-	 *                   .has-<slug>-<class_suffix> {
-	 *                       <property_name>: <preset_value>
-	 *                   }
+	 * - path       => where to find the preset within the settings section
+	 *
+	 * - value_key  => the key that represents the value
+	 *
+	 * - value_func => optionally, instead of value_key, a function to generate
+	 *                 the value that takes a preset as an argument
+	 *
+	 * - css_var    => name of the var to generate. The "$slug" substring will be
+	 *                 replaced by the slug of each preset. For example,
+	 *                 given a preset for color with two values whose slugs are "black" and "white",
+	 *                 the string "--wp--preset--color--$slug" will generate two variables:
+	 *                 "--wp--preset--color--black" and "--wp--preset--color--white".
+	 *
+	 * - classes    => array containing a structure with the classes to
+	 *                 generate for the presets, where for each array item
+	 *                 the key is the class name and the value the property name.
+	 *                 The "$slug" substring will be replaced by the slug of each preset.
+	 *                 For example:
+	 *                 'classes' => array(
+	 *                   '.has-$slug-color'            => 'color',
+	 *                   '.has-$slug-background-color' => 'background-color',
+	 *                   '.has-$slug-border-color'     => 'border-color',
+	 *                 )
```

Comment:
`properties` is missing in the documentation here, but since the purpose of the properties being checked is for the generated `.has-$slug-$property` classes, are `properties` really needed here? 

Duotone, for example, never sets the `filter` property from settings—that's left up to `styles` to handle. We're only setting the `--wp--preset--duotone--$slug` variable from this set of metadata.

Reply:
The `properties` are used exclusively by `remove_insecure_properties` (the kses method) and nowhere else. This method was reusing the `css_var_infix`, but that didn't work for duotone because the infix is not the name of the property to be used (`filter`) but other thing (`duotone`).

Reply:
After this PR, `remove_insecure_properties` understands which property maps to the duotone data (before it tested `duotone: value`, now it tests `filter: value`). The second step it to also teach it that the property value can come from `value_func` (it assumes `value_key` for everything). For simplicity, I'll do that in a follow-up PR.

Reply:
My point is that the CSS that needs to be checked in `remove_insecure_properties` is this for duotone:

```css
--wp--preset--duotone--slug: url('#wp-duotone-slug');
```

It has nothing to do with the `filter` property. That check should be in `remove_insecure_styles`.

And this for things like color:

```css
--wp--preset--color--slug: #000;
.has-slug-color {
  color: var(--wp--preset--color--slug);
}
.has-slug-background-color {
  color: var(--wp--preset--color--slug);
}
.has-slug-border-color {
  color: var(--wp--preset--color--slug);
}
```

So it makes sense to check the `classes` like before—not a new `properties` key.

Reply:
If we forget to update the `properties`, we're not actually testing the generated CSS

Reply:
Ok, I can share a bit about how `remove_insecure_properties` is used, I hope it helps in clarifying things.

That method is used to verify that user data that comes from an import or from an user that doesn't have `unfiltered_html` capabilities is not insecure, according to kses standards. It checks the _values_ of settings (`remove_insecure_settings`) & styles (`remove_insecure_styles`) and removes the ones that are not secure before storing them in the database.

Note that we're not interested in checking the CSS we generate but it checking the _value_ provided _by the user_ is secure. This is an important distinction.

- Let's go a level down into `remove_insecure_settings`. We don't use kses there to validate `--wp--preset--duotone--slug: user_duotone_value` or `--wp--preset--color--slug: user_color_value`: it'd be impossible because kses can't know what's a valid value for any CSS Custom Property. Instead, we use kses to validate that `filter: user_duotone_value` or `color: user_color_value` would be secure according its standards. If these declarations are valid, we can trust the user input. If we can trust the user input, we can use it to generate our own CSS (as custom properties or classes). Now, before introducing `properties` we looked up for the corresponding property by using the affixes we had in `classes` or `css_var_infix` as proxies, because they coincidentally matched the properties to validate. Until we introduced duotone, which didn't match `filter`. By using `properties` we simplify this lookup and decouple the names of the classes/custom properties from the actual property used by kses to validate the user input. Compare the [before](https://github.com/WordPress/gutenberg/blob/771b4564ebfdcdd2913b4e30e8d9d6aeaf607e67/lib/class-wp-theme-json-gutenberg.php#L1301) and [after](https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json-gutenberg.php#L1310). 

- In `remove_insecure_styles` we don't need that substitution because we are already creating the declaration. But the idea is the same, we don't check the block class we generate but only a single declaration (`color: user_color_value`).

Reply:
It's more like this. `remove_insecure_properties` takes as input the following theme.json coming from the user:

```
{
  "settings": {
    "color": {
      "customColor": false, // not allowed
      "palette": [
        {
          "name": "Pink",
          "slug": "pink",
          "color": "#FFC0CB"
        },
        {
          "name": "Red/><b>ok</ok>", // unsafe
          "slug": "red",
          "color": "#ff0000"
        },

        {
          "name": "Blue",
          "slug": "blue",
          "color": "var(--color, var(--unsafe-fallback))" // unsafe
        }
      ]
    }
  },
  "styles": {
    "color": {
      "gradient": "url(data:...)", // unsafe
      "text": "var:preset|color|dark-red"
    },
    "blocks": {
      "core/image": {
        "filter": {
          "duotone": "var:preset|duotone|blue-red"
        }
      }
    }
  }
}
```

And the output (what we save in the database) is going to be (note that the things that were unsafe or not allowed are removed):

```
{
  "settings": {
    "color": {
      "palette": [
        {
          "name": "Pink",
          "slug": "pink",
          "color": "#FFC0CB"
        }
      ]
    }
  },
  "styles": {
    "color": {
      "text": "var:preset|color|dark-red"
    },
    "blocks": {
      "core/image": {
        "filter": {
          "duotone": "var:preset|duotone|blue-red"
        }
      }
    }
  }
}
```

What happens in between is that:

- For styles, we check that every value is safe according to kses
- For settings, we check that every preset is safe according to kses. This means names (which are going to be used in the UI), slugs (used to generate classes & custom properties), and values (same as before).

For checking the value we use kses as a intermediate tool (`safecss_filter_attr`). In the case of settings, kses wouldn't know how to validate `--wp--preset--color--slug: user_color_value` hence why we pass it `color: user_color_value` instead. This substitution is something we do only for checking the user value, it's not used anywhere else.

Because a custom property like `--wp--preset--color--slug` can be used in different situations (`color: var( --wp...)`, `background-color: var(--wp..)`, `border-color: var(--wp...)`) we check all the properties that apply for a certain preset. If any of them fails, the value will be removed. Arguably, in certain situations we could use a single property (`color` instead of `[ color, background-color, border-color]` but that presumes some knowledge about how kses works internally, which should be transparent to us.

This is the reason I believe what we have now (`properties` instead of checking affixes in `classes` and `css_vars`) is a lot clearer than what we have before.

Does this help? I'm happy to jump into a call to go through this as well.

Reply:
By the way, https://github.com/WordPress/gutenberg/pull/35255 is the continuation of this work. It updates `remove_insecure_settings` to consider presets that use the `value_func`. Welcome reviews :)

Reply:
https://github.com/WordPress/gutenberg/pull/35318 is for making `remove_insecure_styles` work with duotone data.

Reply:
Thanks for the well-written explanation! I've also stepped through `remove_insecure_properties` and into `safecss_filter_attr` in a debugger to get a more detailed idea of exactly what is going on.

> In the case of settings, kses wouldn't know how to validate `--wp--preset--color--slug: user_color_value` hence why we pass it `color: user_color_value` instead.

This is exactly what feels weird to me that I was trying to get at. Maybe `safecss_filter_attr` isn't the right tool for what we're trying to do. A function that is aware of the CSS custom properties and doesn't have to substitute CSS properties where the variable _might_ be used would be more robust and less confusing. 

> Arguably, in certain situations we could use a single property (`color` instead of `[ color, background-color, border-color]`) but that presumes some knowledge about how kses works internally, which should be transparent to us.

We have to add properties _because_ we know how kses (`safecss_filter_attr`) works internally—it can't handle CSS custom properties. If we wanted to be truly transparent to how kses works, we should be able to pass the properties directly to either kses or a new function that can handle them as they appear in the stylesheet.

> Does this help? I'm happy to jump into a call to go through this as well.

If you want to talk more about it, I'd be happy to hop on a call and write up a summary for here. I might not be explaining my perspective on this very well. 😅

Reply:
> We have to add properties in the first place because we know how kses (safecss_filter_attr) works internally

Not really! We add properties because that's how the CSS vars are going to be used in the classes & styles the engine generates. For example, color presets generate the vars plus the classes:

```
.has-slug-color { COLOR: <our-css-var>; }
.has-slug-background-color { BACKGROUND-COLOR: <our-css-var>; }
.has-slug-border-color { BORDER-COLOR: <our-css-var>; }
```

And they are also expected to be used in the corresponding `theme.json` styles: `styles.border.color`, `styles.color.background`, `styles.color.text`, etc.

---

We could argue that this info is already in the `classes` key of the `presets_metadata` array. However, not all presets generate classes (e.g. duotone), hence why we need a separate `properties` key for clarity.

Reply:
> We could argue that this info is already in the classes key of the presets_metadata array. However, not all presets generate classes (e.g. duotone), hence why we need a separate properties key for clarity.

Couldn't `remove_insecure_properties` keep track of the CSS custom properties that we use and check them without needing to manually add the `properties` key?